contracts/RefundProvider/RefundProvider.sol

```diff
@@ -0,0 +1,115 @@
+        uint256 [] memory mainCoinParams = new uint256[](2);
+        mainCoinParams[0] = params[paramsLength - 2];
+        mainCoinParams[1] = params[paramsLength - 1];
+        uint256 lockProviderPoolId = lockDealNFT.mint(msg.sender, mainCoin, msg.sender, mainCoinParams[0], address(lockProvider));
+        lockProvider.registerPool(lockProviderPoolId, mainCoinParams);
+
+        /// real owner poolId
+        poolId = lockDealNFT.mint(owner, token, msg.sender, 0, address(this));
+    }
+
+    ///@dev split tokens and main coins into new pools
+    function split(
+        uint256 poolId,
+        uint256,
+        uint256 splitAmount
+    ) external onlyNFT {
+        address provider = lockDealNFT.poolIdToProvider(poolId - 2);
+        uint256 tokenLeftAmount = lockProvider.dealProvider().poolIdToleftAmount(poolId - 2);
+        address mainCoin = lockDealNFT.tokenOf(poolId - 1);
+        uint256 mainCoinAmount = lockProvider.dealProvider().poolIdToleftAmount(poolId - 1);
+        uint256 mainCoinSplitAmount = _calcAmount(splitAmount, _calcRate(tokenLeftAmount, mainCoinAmount));
+
+        uint256 newPoolId = lockDealNFT.mint(lockDealNFT.ownerOf(poolId - 2), lockDealNFT.tokenOf(poolId - 2), msg.sender, 0, provider);
+        IProvider(provider).split(poolId - 2, newPoolId, splitAmount);
+        lockDealNFT.setPoolIdToVaultId(newPoolId, lockDealNFT.poolIdToVaultId(poolId - 2));
+
+        uint256 lockProviderPoolId = lockDealNFT.mint(lockDealNFT.ownerOf(poolId - 1), mainCoin, msg.sender, 0, address(lockProvider));
+        lockProvider.split(poolId - 1, lockProviderPoolId, mainCoinSplitAmount);
+        lockDealNFT.setPoolIdToVaultId(lockProviderPoolId, lockDealNFT.poolIdToVaultId(poolId - 1));
+
+        lockDealNFT.mint(lockDealNFT.ownerOf(poolId), lockDealNFT.tokenOf(poolId), msg.sender, 0, address(this));
+    }
+
+    function _calcRate(uint256 tokenAValue, uint256 tokenBValue) internal pure returns (uint256) {
+        return tokenBValue != 0 ? (tokenAValue * 1e18) / tokenBValue : 0;
+    }
+
+    function _calcAmount(uint256 amount, uint256 rate) internal pure returns (uint256) {
+        return rate != 0 ? (amount * 1e18) / rate : 0;
+    }
+
+    function withdraw(
+        uint256 poolId
+    ) public override onlyNFT returns (uint256 withdrawnAmount, bool isFinal) {
+        // user withdraws his token from the provider by cascading providers params
+        DealProvider dealProvider = lockProvider.dealProvider();
+        uint256 tokenAmount = dealProvider.poolIdToleftAmount(poolId - 2);
+        (withdrawnAmount, isFinal) = lockDealNFT.withdraw(poolId - 2);
+
+        if (withdrawnAmount > 0 && lockProvider.startTimes(poolId - 1) <= block.timestamp) {
+            address mainCoin = lockDealNFT.tokenOf(poolId - 1);
+            uint256 mainCoinAmount = lockProvider.dealProvider().poolIdToleftAmount(poolId - 1);
+            uint256 withdrawMainCoinAmount = _calcAmount(withdrawnAmount, _calcRate(tokenAmount, mainCoinAmount));
```

Comment:
I think `_calcRate()` will always returns 100%

Reply:
fixed by: https://github.com/The-Poolz/LockDealNFT/blob/a711c7d4e73454f00017fab684d35df1031cf061/contracts/RefundProvider/RefundProvider.sol#L105